### PR TITLE
Public FT serializer

### LINKED ISSUE #1652: Have a common public FormulaType serializer.
From: https://dynamicscrm.visualstudio.com/OneCRM/_workitems/edit/3437244/
Needed by DV Plugin.
Today we have 2 internal FT serializers. We should have one common public one.

Obsolete Serializer: https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.LanguageServerProtocol/JsonRpc/ObsoleteTypeSerializer/FormulaTypeJsonConverter.cs

Non-Obsolete Serializer: https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeJsonConverter.cs

**Note:** Both serializers are serializing differently and are not compatible with each other.

### PR DESCRIPTION
**Pull request by jas-valgotar**
Fixes #1652

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeJsonConverter.cs [hunk 1]
```diff
@@ -12,24 +12,37 @@
 
 namespace Microsoft.PowerFx.Core
 {
-    internal class FormulaTypeJsonConverter : JsonConverter<FormulaType>
+    public class FormulaTypeJsonConverter : JsonConverter<FormulaType>
     {
         private readonly DefinedTypeSymbolTable _definedTypes;
 
-        public FormulaTypeJsonConverter(DefinedTypeSymbolTable definedTypes)
+        private Func<string, RecordType> _logicalNameToRecordType;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeJsonConverter.cs:19]: >private [](http://example.com/codeflow?start=8&length=7)

nit : readonly?

Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeJsonConverter.cs]: Updated

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeJsonConverter.cs [hunk 2]
```diff
@@ -12,24 +12,37 @@
...
     {
         private readonly DefinedTypeSymbolTable _definedTypes;
 
-        public FormulaTypeJsonConverter(DefinedTypeSymbolTable definedTypes)
+        private readonly Func<string, RecordType> _logicalNameToRecordType;
+
+        internal FormulaTypeJsonConverter(DefinedTypeSymbolTable definedTypes)
         {
             _definedTypes = definedTypes;
         }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeJsonConverter.cs:24]: _logicalNameToRecordType will be null in this ctor

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSchema.cs [hunk 3]
```diff
@@ -54,18 +55,28 @@ public FormulaType ToFormulaType(DefinedTypeSymbolTable definedTypeSymbols)
...
-                return FormulaType.BindingError;
+                return logicalNameToRecordType.Invoke(this.Description).ToTable();
             }
-
-            if (Fields == null || !Fields.Any())
+            else if (typeName == SchemaTypeName.ExpandableRecordTypeName.Name && logicalNameToRecordType != null)
+            {
+                return logicalNameToRecordType.Invoke(this.Description);
+            }
+            else if (typeName == SchemaTypeName.RecordTypeName.Name)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSchema.cs:67]: IMO, we shouldn't be using "ExpandTable" + the description to handle serializing the types of related entities. 

Instead, consider adding those types to a definedTypeSymbolTable, and using the type names (which, in discussion with Greg when doing the Custom Page work, we determined would be the registered table names). 

i.e. 

a table *[Foo: E] where E is to the Accounts table and is tabular, should be serialized as 

```
{
    Type: {
         TypeName: Record
          IsTable: true
     }
      Fields: 
      {
             Foo:
             {
                 Type: {
                      TypeName: Accounts
                      IsTable: true
                   }
             }
      }
}
````

File: src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSchema.cs [hunk 4]
```diff
@@ -120,6 +131,10 @@ public override int GetHashCode()
 
         public static SchemaTypeName RecordTypeName => new () { Name = "Record", IsTable = false };
 
+        public static SchemaTypeName ExpandableTableTypeName => new () { Name = "ExpandableTable", IsTable = true };
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSchema.cs:134]: >ExpandableTable [](http://example.com/codeflow?start=81&length=15)

The point of `IsTable` is that that's what distinguishes table. This should just be `ExpandableRecord` with IsTable == true.

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSchema.cs:134]: see XR doc

File: src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSerializerSerttings.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core
+{
+    public class FormulaTypeSerializerSerttings
```
[the remark this answers is not in the document]
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSerializerSerttings.cs:11]: #1736

File: src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSchema.cs (hunk 3 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Json/Public/Types/UserProvidedTypeDefinitions/FormulaTypeSchema.cs:67]: Discussed offline.

**Pull request merged**